packages/edit-site/src/components/navigation-sidebar/navigation-panel/menus/templates.js

```diff
@@ -50,26 +67,45 @@ export default function TemplatesMenu() {
 			title={ __( 'Templates' ) }
 			titleAction={ <NewTemplateDropdown /> }
 			parentMenu={ MENU_ROOT }
+			hasSearch={ true }
+			onSearch={ onSearch }
+			search={ search }
 		>
-			<NavigationItem
-				navigateToMenu={ MENU_TEMPLATES_ALL }
-				title={ _x( 'All', 'all templates' ) }
-			/>
-			<NavigationItem
-				navigateToMenu={ MENU_TEMPLATES_PAGES }
-				title={ __( 'Pages' ) }
-			/>
-			<NavigationItem
-				navigateToMenu={ MENU_TEMPLATES_POSTS }
-				title={ __( 'Posts' ) }
-			/>
-
-			{ map( generalTemplates, ( template ) => (
-				<TemplateNavigationItem
-					item={ template }
-					key={ `wp_template-${ template.id }` }
+			{ search && (
+				<SearchResults
+					items={ templates }
+					searchQuery={ searchQuery }
+					renderItem={ renderSearchResultItem }
+					isDebouncing={ isDebouncing }
 				/>
-			) ) }
+			) }
+
+			{ ! search && (
+				<>
+					<NavigationItem
+						navigateToMenu={ MENU_TEMPLATES_ALL }
+						title={ _x( 'All', 'all templates' ) }
+					/>
+					<NavigationItem
+						navigateToMenu={ MENU_TEMPLATES_PAGES }
+						title={ __( 'Pages' ) }
+					/>
+					<NavigationItem
+						navigateToMenu={ MENU_TEMPLATES_POSTS }
+						title={ __( 'Posts' ) }
+					/>
+					{ map( generalTemplates, ( template ) => (
+						<TemplateNavigationItem
+							item={ template }
+							key={ `wp_template-${ template.id }` }
+						/>
+					) ) }
+				</>
+			) }
+
+			{ ! search && templates === null && (
+				<NavigationItemText title={ __( 'Loading…' ) } />
+			) }
```

Comment:
This (and its template parts counterpart) creates a curious side effect when opening the sidebar for the first time:

<img width="280" alt="Screenshot 2020-11-11 at 14 23 17" src="https://user-images.githubusercontent.com/2070010/98824245-fc985480-242a-11eb-9172-bc30bd8f94ec.png">

I would recommend to keep these as simple `NavigationItem` for now, and maybe look into it separately.

For example, we could address it by moving the `Templates >` item into this same component, and render that or "Loading" alternatively. E.g.
```jsx
<>
  { templates && <NavigationItem navigateToMenu="templates" /> }
  { ! search && ! templates && <NavigationItemText title="Loading" /> }
  { <NavigationMenu menu="templates" /> }
<>
```

(Just to be clear, I'm only talking about the "loading" item **not** part of the search.)

Reply:
Gotcha, this happens because the custom "NavigationItemText" doesn't register into the navigation tree. So it is always visible.

I'm spinning up a PR to introduce `NavigationItemText` as a component in the `navigation` package. This will be solved by using the built-in component rather than introducing it here.